File: java/src/main/java/core/Partition.java
package core;

import java.util.Arrays;
import java.util.Iterator;
import java.util.Random;

import scala.Tuple2;

import org.apache.spark.HashPartitioner;
import org.apache.spark.Partitioner;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.VoidFunction;

public class Partition {

    public static void main(String[] args) {
        SparkConf conf = new SparkConf().setAppName("Partition");
        JavaSparkContext sc = new JavaSparkContext(conf);

        @SuppressWarnings("unchecked")
        JavaPairRDD<Integer, String> nums1 = sc.parallelizePairs(Arrays.asList(
            new Tuple2<Integer, String>(1, "a"),
            new Tuple2<Integer, String>(2, "b"),
            new Tuple2<Integer, String>(3, "c"),
            new Tuple2<Integer, String>(4, "d"),
            new Tuple2<Integer, String>(5, "e")
        ));
        JavaPairRDD<Integer, String> nums2 = nums1.partitionBy(new HashPartitioner(2));
        JavaPairRDD<Integer, String> nums3 = nums1.partitionBy(new RandomPartitioner(2));

        class PrintNums implements VoidFunction<Iterator<Tuple2<Integer, String>>> {
            public void call(Iterator<Tuple2<Integer, String>> it) throws Exception {
                while (it.hasNext()) {
                    Tuple2<Integer, String> e = it.next();
                    System.out.print(String.format("(%d -> %s) ", e._1(), e._2()));
                }
                System.out.println();
            }
        }

        System.out.println("### Default ###");
        nums1.foreachPartition(new PrintNums());
        System.out.println();

        System.out.println("### HashPartitioner ###");
        nums2.foreachPartition(new PrintNums());
        System.out.println();

        System.out.println("### RandomPartitioner ###");
        nums3.foreachPartition(new PrintNums());
        System.out.println();
    }

    public static class RandomPartitioner extends Partitioner {
        private int numParts;
        private Random rnd = new Random();

        public RandomPartitioner(int numParts) {
            this.numParts = numParts;
        }

        public int numPartitions() {
            return numParts;
        }

        public int getPartition(Object key) {
            return rnd.nextInt(Integer.MAX_VALUE) % numPartitions();
        }

        public boolean equals(Object other) {
            if (other == this) {
                return true;
            }
            if (!(other instanceof RandomPartitioner)) {
                return false;
            }
            return ((RandomPartitioner) other).numPartitions() == numPartitions();
        }
    }
}

File: scala/src/main/scala/core/Partition.scala
package core


import java.util.Random

import org.apache.spark.SparkContext._
import org.apache.spark._

object Partition {

  def main(args: Array[String]) {
    run()
  }

  def run() {
    val conf = new SparkConf().setAppName("Partition")
    val sc = new SparkContext(conf)

    val nums1 = sc.parallelize(Array((1, "a"), (2, "b"), (3, "c"), (4, "d"), (5, "e"))).persist()
    val nums2 = nums1.partitionBy(new HashPartitioner(2))
    val nums3 = nums1.partitionBy(new RangePartitioner(3, nums1))
    val nums4 = nums1.partitionBy(new RandomPartitioner(2))

    println("### Default ###")
    nums1.foreachPartition(p => println(p.toSeq.mkString(",")))
    println()

    println("### HashPartitioner ###")
    nums2.foreachPartition(p => println(p.toSeq.mkString(",")))
    println()

    println("### RangePartitioner ###")
    nums3.foreachPartition(p => println(p.toSeq.mkString(",")))
    println()

    println("### RandomPartitioner ###")
    nums4.foreachPartition(p => println(p.toSeq.mkString(",")))
    println()
  }
}

class RandomPartitioner(numParts: Int) extends Partitioner with Serializable {
  val rnd = new Random()

  override def numPartitions: Int = numParts

  override def getPartition(key: Any): Int = {
    rnd.nextInt(Int.MaxValue) % numPartitions
  }

  override def equals(other: Any): Boolean = {
    other match {
      case p: RandomPartitioner => p.numPartitions == numPartitions
      case _ => false
    }
  }
}
